pkg/kubelet/nodeconfig/checkpoint.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeconfig
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	kuberuntime "k8s.io/apimachinery/pkg/runtime"
+	api "k8s.io/kubernetes/pkg/api"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+)
+
+// checkpointExists returns true if checkpoint for the config source identified by `uid` exists on disk.
+// If the existence of a checkpoint cannot be determined due to filesystem issues, a fatal error occurs.
+func (cc *NodeConfigController) checkpointExists(uid string) bool {
+	ok, err := cc.dirExists(filepath.Join(checkpointsDir, uid))
+	if err != nil {
+		fatalf("failed to determine whether checkpoint %q exists, error: %v", uid, err)
+	}
+	return ok
+}
+
+// loadCheckpoint loads the checkpoint at `cc.configDir/relPath`, which is expected to be a checkpoint directory.
+// If the checkpoint directory does not exist or if data cannot be retrieved from the filesystem,
+// a fatal error will occur.
+// If the data cannot be decoded and converted to a supported config source type, returns an error.
+// This may indicate a failure to completely save the checkpoint. You may want to attempt a re-download in this scenario.
+// If loading succeeds, returns a `verifiable` (see verify.go). This interface can be used to verify the integrity of
+// the loaded checkpoint.
+func (cc *NodeConfigController) loadCheckpoint(relPath string) (verifiable, error) {
+	path := filepath.Join(cc.configDir, relPath)
+	infof("loading configuration from %q", path)
+
+	// find the checkpoint file(s)
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		fatalf("failed to enumerate checkpoint files in dir %q, error: %v", path, err)
+	} else if len(files) == 0 {
+		return nil, fmt.Errorf("no checkpoint files in dir %q, but there should be at least one", path)
+	}
+
+	// TODO(mtaufen): for now, we only have one file per checkpoint (a serialized API object, e.g. a ConfigMap); if this ever changes we will need to extend this
```

Comment:
It's less error prone (and less code) to just checkpoint the wrapper object, rather than try to unpack into several files, and then load from these later. 

Checkpointing the wrapper object also has the nice side-benefit of retaining the ObjectMeta of the config source (hash is in the name), which means you can verify the checkpoint whenever you want without having to have stored the hash somewhere else when initially saving it.